feat: add Pytorch support of Vision Encoder for multimodal models

qixiang-99 opened the pull request:
# feat: add Pytorch support of Vision Encoder for multimodal models

This PR supports Siglip and Clip vision model for multimodal models. 

## Description
The changes in this PR can be categorized into these sections:
### Model Support
**Note: Please note that in this PR, Siglip and Clip vision model are tailored for multimodal models as the visual encoder.**
1.  Support Clip vision model
2. Support Siglip vision model
3. Integrate TRT-LLM Clip vision model into existing llava-next model

#### Clip Vision Model
Here is the components break down:
| Componenet           | Module Class        |
|----------------------|---------------------|
| Self Attn            | TRTLLM              |
| Layer Norm           | torch.nn.<br>Module |
| MLP                  | TRTLLM              |
| CLIPVisionEmbeddings | HF                  |

#### Siglip Vision Model
Here is the components break down:
| Componenet           | Module Class        |
|----------------------|---------------------|
| Self Attn            | TRTLLM              |
| Layer Norm           | torch.nn.<br>Module |
| MLP                  | TRTLLM              |
| SiglipVisionEmbeddings | HF                  |

### 

## Test Coverage

<!--
Please list clearly what are the relevant test(s) that can safeguard the changes in the PR. This helps us to ensure we have sufficient test coverage for the PR.
-->
### Unit Test
`tests/unittest/_torch/modeling/test_modeling_clip.py`
`tests/unittest/_torch/modeling/test_modeling_siglip.py`
### Integration Test
`tests/integration/defs/test_e2e.py::test_ptp_quickstart_multimodal[llava-v1.6-mistral-7b-llava-v1.6-mistral-7b-hf-image]`

## GitHub Bot Help

`/bot [-h] ['run', 'kill', 'skip', 'reuse-pipeline'] ...`

Provide a user friendly way for developers to interact with a Jenkins server.

Run `/bot [-h|--help]` to print this help message.

See details below for each supported subcommand.

<details>

`run  [--disable-fail-fast --skip-test --stage-list "A10-1, xxx" --gpu-type "A30, H100_PCIe" --add-multi-gpu-test --only-multi-gpu-test --disable-multi-gpu-test --post-merge --extra-stage "H100_PCIe-[Post-Merge]-1, xxx"]`

Launch build/test pipelines. All previously running jobs will be killed.

`--disable-fail-fast ` *(OPTIONAL)* : Disable fail fast on build/tests/infra failures.

`--skip-test ` *(OPTIONAL)* : Skip all test stages, but still run build stages, package stages and sanity check stages. Note: Does **NOT** update GitHub check status.

`--stage-list "A10-1, xxx"` *(OPTIONAL)* : Only run the specified test stages. Examples: "A10-1, xxx". Note: Does **NOT** update GitHub check status.

`--gpu-type "A30, H100_PCIe"` *(OPTIONAL)* : Only run the test stages on the specified GPU types. Examples: "A30, H100_PCIe". Note: Does **NOT** update GitHub check status.

`--only-multi-gpu-test ` *(OPTIONAL)* : Only run the multi-GPU tests. Note: Does **NOT** update GitHub check status.

`--disable-multi-gpu-test ` *(OPTIONAL)* : Disable the multi-GPU tests. Note: Does **NOT** update GitHub check status.

`--add-multi-gpu-test ` *(OPTIONAL)* : Force run the multi-GPU tests. Will also run L0 pre-merge pipeline.

`--post-merge ` *(OPTIONAL)* : Run the L0 post-merge pipeline instead of the ordinary L0 pre-merge pipeline.

`--extra-stage "H100_PCIe-[Post-Merge]-1, xxx"` *(OPTIONAL)* : Run the ordinary L0 pre-merge pipeline and specified test stages. Examples: --extra-stage "H100_PCIe-[Post-Merge]-1, xxx".

### kill

`kill  `

Kill all running builds associated with pull request.

### skip

`skip --comment COMMENT `

Skip testing for latest commit on pull request. `--comment "Reason for skipping build/test"` is required. IMPORTANT NOTE: This is dangerous since lack of user care and validation can cause top of tree to break.

### reuse-pipeline

`reuse-pipeline `

Reuse a previous pipeline to validate current commit. This action will also kill all currently running builds associated with the pull request. IMPORTANT NOTE: This is dangerous since lack of user care and validation can cause top of tree to break.

</details>

qixiang-99 commented:
/bot run --disable-fail-fast

tensorrt-cicd commented:
[PR_Github #3140](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/3140/) [ run ] triggered by Bot

tensorrt-cicd commented:
[PR_Github #3140](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/3140/) [ run ] completed with state `SUCCESS`
[/LLM/main/L0_MergeRequest_PR pipeline #2188](https://nv/trt-llm-cicd/job/main/job/L0_MergeRequest_PR/2188/) completed with status: 'SUCCESS'

symphonylyh commented on tensorrt_llm/_torch/models/modeling_siglip.py:
```diff
@@ -0,0 +1,303 @@
+            'gate_up_proj': ['gate_proj', 'up_proj']
+        }
+        for name, module in tqdm(list(self.named_modules()),
+                                 desc="Loading weights"):
+            if len(module._parameters) > 0:
+                # skip load weights if tie word embeddings is enabled and layer is lm_head
+                if self.config.tie_word_embeddings and name.startswith(
+                        "lm_head"):
+                    continue
+
+                # Skip if parameter belongs to a missing layer
+                if missing_layer_parameter(name, self):
+                    continue
+
+                names = name.split('.')
+                if names[-1] in params_map:
+                    module_weights = []
+                    for new_name in params_map[names[-1]]:
+                        fw = filter_weights('.'.join(names[:-1] + [new_name]),
+                                            weights)
+                        if new_name in ['k_proj', 'v_proj']:
+                            fw = {
+                                k:
+                                duplicate_kv_weight(
+                                    weight=v[:],
+                                    head_dim=head_dim,
+                                    tensor_parallel_size=tp_size)
+                                if k in ["weight", "bias"] else v
+                                for k, v in fw.items()
+                            }
+                        module_weights.append(fw)
+                    module.load_weights(weights=module_weights)
+                else:
+                    module_weights = filter_weights(name, weights)
+                    if hasattr(module, 'load_weights'):
+                        module.load_weights(weights=[module_weights])
+                    else:
+                        for n, p in module._parameters.items():
+                            if p is not None:
+                                p.data.copy_(module_weights[n][:])
```
is this identical to the one in modeling_utils.py? this part we can refactor it to be a function in modeling_utils.py and let DecoderModelForCausualLM call it, meanwhile other modeling_x.py can also from .modeling_utils import load_weights_generic to avoid duplication

qixiang-99 replied:
Refactored [here](https://github.com/NVIDIA/TensorRT-LLM/pull/3791/files#diff-1f75fa2771954bee99cca07b03d7b5c30605e2c47e5ce1d61405424023b13361R617-R691) with optional regex weight mapping

qixiang-99 commented on examples/pytorch/quickstart_multimodal.py:
```diff
@@ -87,6 +87,9 @@ def parse_arguments():
 
 def main():
     args = parse_arguments()
+    # set prompts and media to example prompts and images if they are not provided
+    args.prompt = example_image_prompts if args.modality == "image" else example_video_prompts
+    args.media = example_images if args.modality == "image" else example_videos
```
Yes this is wrong. Will change it when I rebase.

qixiang-99 commented on tensorrt_llm/_torch/models/modeling_llava_next.py:
```diff
@@ -25,10 +27,16 @@ def __init__(self, model_path, model_config, tokenizer):
-            model_path, torch_dtype=model_config.text_config.torch_dtype)
         self.device = 'cuda'
-        self.vision_tower = model.vision_tower.vision_model.to(self.device)
+        model = LlavaNextForConditionalGeneration.from_pretrained(
+            model_path,
+            torch_dtype=model_config.text_config.torch_dtype).to(self.device)
+        vision_model_config = ModelConfig(
+            pretrained_config=model_config.vision_config, attn_backend="TRTLLM")
+        self.vision_tower = CLIPVisionModel(vision_model_config).to(
+            self.device).to(model.dtype)
+        self.vision_tower.load_weights(model.vision_tower.state_dict())
+
         self.mm_projector = model.multi_modal_projector.to(self.device)
```
Yes, I have the same idea here but didn't be aware of the `load_sharded_checkpoint` method. I think this is doable with this method. Will try, thanks for suggestion!

qixiang-99 replied:
Updated

qixiang-99 committed:
feat: Add rename_weights_with_regex function for dynamic weight key renaming

Introduced a new utility function to rename weight keys in a dictionary based on regex pattern matching. This allows for flexible mapping of keys from Hugging Face naming conventions to TRT-LLM naming conventions, enhancing model compatibility and usability.

Signed-off-by: qixiang-99 <203170375+qixiang-99@users.noreply.github.com>

qixiang-99 committed:
feat: Implement SiglipVisionModel and related components

Added the SiglipVisionModel along with its associated classes, including SiglipAttention, SiglipEncoderLayer, and SiglipEncoder.
Additionally, a new test suite for the SiglipVisionModel has been created to ensure compatibility with Hugging Face outputs.

Currently SiglipVisionModel support batch size larger than one. Also, inputs and outputs shape are same with the HF for compatibility.

Signed-off-by: qixiang-99 <203170375+qixiang-99@users.noreply.github.com>

qixiang-99 committed:
feat: Add CLIPVisionModel and associated components

Introduced the CLIPVisionModel along with its related classes, including CLIPAttention, CLIPEncoderLayer, CLIPEncoder, and CLIPVisionTransformer. This implementation aligns with Hugging Face's CLIP architecture, ensuring compatibility in input and output shapes.

Signed-off-by: qixiang-99 <203170375+qixiang-99@users.noreply.github.com>

qixiang-99 committed:
feat: Enhance CLIPVisionModel with attention metadata preparation and unit tests

Updated the CLIPVisionModel to include a method for preparing attention metadata, simplifying the model's usage. Additionally, added a comprehensive unit test suite for the CLIPVisionModel, ensuring compatibility with Hugging Face outputs and validating model performance across various scenarios.

Signed-off-by: qixiang-99 <203170375+qixiang-99@users.noreply.github.com>

qixiang-99 committed:
feat: Refactor SiglipVisionModel with attention metadata preparation and update unit tests

Enhanced the SiglipVisionModel by adding a method to prepare attention metadata, streamlining its usage. Updated unit tests to validate model performance and compatibility with Hugging Face outputs, including adjustments to the configuration and test scenarios.

Signed-off-by: qixiang-99 <203170375+qixiang-99@users.noreply.github.com>

qixiang-99 committed:
refactor: Remove unused rotary_emb parameter from CLIP and Siglip attention classes

Eliminated the rotary_emb parameter from the CLIPAttention and SiglipAttention classes to streamline the code. Updated unit tests to reflect changes in the model configurations, including clarifications in the default configurations sourced from Hugging Face.

Signed-off-by: qixiang-99 <203170375+qixiang-99@users.noreply.github.com>

qixiang-99 committed:
feat: Integrate CLIPVisionModel into LlavaNextInputProcessor and enhance weight loading

Added CLIPVisionModel to the LlavaNextInputProcessor for improved vision processing. Updated the model loading mechanism to ensure compatibility with the new vision model and added attention metadata preparation. Removed debug print statements from weight renaming function for cleaner code.

Signed-off-by: qixiang-99 <203170375+qixiang-99@users.noreply.github.com>

qixiang-99 committed:
refactor: Remove unused max_position_embeddings from CLIPAttention and update Siglip classes to use CLIP components

Removed the unused max_position_embeddings variable from the CLIPAttention class. Updated the Siglip classes to utilize CLIP components, specifically replacing SiglipEncoder and SiglipAttention with their CLIP counterparts, streamlining the codebase and enhancing consistency across models.

Signed-off-by: qixiang-99 <203170375+qixiang-99@users.noreply.github.com>

qixiang-99 committed:
refactor: Consolidate weight loading logic into a shared implementation

Refactored the weight loading process across CLIP and Siglip models by using a new utility function, _load_weights_impl, to streamline the loading mechanism. This change enhances code maintainability and reduces redundancy in weight handling, ensuring consistent behavior across different model architectures.

Signed-off-by: qixiang-99 <203170375+qixiang-99@users.noreply.github.com>

qixiang-99 committed:
refactor: Simplify output handling in CLIP and Siglip models by removing output_hidden_states parameter

Removed the output_hidden_states parameter from the CLIPEncoder and SiglipVisionTransformer classes, streamlining the output handling process. Updated the corresponding unit tests to reflect these changes and ensure compatibility with the new output structure.

Signed-off-by: qixiang-99 <203170375+qixiang-99@users.noreply.github.com>

qixiang-99 committed:
feat: Enhance LlavaNextInputProcessor with dynamic model loading and memory optimization

Updated the LlavaNextInputProcessor to support dynamic model loading from local paths or Hugging Face, improving memory efficiency by partially loading the model components. Integrated the LlavaNextMultiModalProjector and adjusted weight loading to ensure compatibility with the new architecture.

Signed-off-by: qixiang-99 <203170375+qixiang-99@users.noreply.github.com>

qixiang-99 commented:
/bot run --disable-fail-fast

tensorrt-cicd commented:
[PR_Github #3895](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/3895/) [ run ] completed with state `SUCCESS`
[/LLM/main/L0_MergeRequest_PR pipeline #2762](https://nv/trt-llm-cicd/job/main/job/L0_MergeRequest_PR/2762/) completed with status: 'SUCCESS'

qixiang-99 commented on tensorrt_llm/_torch/models/modeling_siglip.py:
```diff
@@ -0,0 +1,303 @@
+        self.config = config
+        embed_dim = config.hidden_size
+
+        self.embeddings = SiglipVisionEmbeddings(config)
```
Documented in PR description.

qixiang-99 commented on tensorrt_llm/_torch/models/modeling_siglip.py:
Updated.

symphonylyh commented on tensorrt_llm/_torch/models/modeling_llava_next.py:
```diff
@@ -25,11 +32,40 @@ def __init__(self, model_path, model_config, tokenizer):
+        self.vision_tower.load_weights(hf_vision_tower.state_dict())
+
+        # Use HF multi-modal projector
+        self.mm_projector = hf_mm_projector
```
it's ok to leave mm projector as HF for now. but when you do the nsys profiling to show TRT-LLM ViT vs. HF ViT, it would be good to record the latency of mm projector too (I think it's just one gemm)

qixiang-99 commented on tensorrt_llm/_torch/models/modeling_clip.py:
As discussed offline, enabling TP is straightforward—add the mapping configuration in the `ModelConfig` class so we can support TP in Attention and MLP

qixiang-99 commented:
/bot run --disable-fail-fast

tensorrt-cicd commented:
[PR_Github #3981](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/3981/) [ run ] triggered by Bot

tensorrt-cicd commented:
[PR_Github #3981](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/3981/) [ run ] completed with state `SUCCESS`
[/LLM/main/L0_MergeRequest_PR pipeline #2824](https://nv/trt-llm-cicd/job/main/job/L0_MergeRequest_PR/2824/) completed with status: 'SUCCESS'

symphonylyh committed:
Merge branch 'main' into feat/vit-models

tensorrt-cicd commented:
[PR_Github #3990](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/3990/) [ reuse-pipeline ] triggered by Bot

tensorrt-cicd commented:
[PR_Github #3990](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/3990/) [ reuse-pipeline ] completed with state `SUCCESS`
Reusing [PR_Github #3981](https://prod.blsm.nvidia.com/sw-tensorrt-top-1/job/LLM/job/helpers/job/PR_Github/3981/) for commit `0842053`
